static/js/presence.js

```diff
@@ -41,53 +41,108 @@ exports.get_user_ids = function () {
     return Array.from(exports.presence_info.keys());
 };
 
-function status_from_timestamp(baseline_time, info) {
-    let status = 'offline';
-    let last_active = 0;
-
-    for (const [device, device_presence] of Object.entries(info)) {
-        const age = baseline_time - device_presence.timestamp;
-        if (last_active < device_presence.timestamp) {
-            last_active = device_presence.timestamp;
+exports.status_from_raw = function (raw) {
+    /*
+        Example of `raw`:
+
+        {
+            active_timestamp: 1585745133
+            idle_timestamp: 1585745091
+            server_timestamp: 1585745140
         }
-        if (age < OFFLINE_THRESHOLD_SECS) {
-            switch (device_presence.status) {
-            case 'active':
-                status = device_presence.status;
-                break;
-            case 'idle':
-                if (status !== 'active') {
-                    status = device_presence.status;
-                }
-                break;
-            case 'offline':
-                if (status !== 'active' && status !== 'idle') {
-                    status = device_presence.status;
-                }
-                break;
-            default:
-                blueslip.error('Unexpected status', {presence_object: device_presence, device: device}, undefined);
+    */
+    function age(timestamp) {
+        return raw.server_timestamp - (timestamp || 0);
+    }
+
+    const active_timestamp = raw.active_timestamp;
+    const idle_timestamp = raw.idle_timestamp;
+
+    /*
+        If the server sends us `active_timestamp`, this
+        means at least one client was active at this time
+        (and hasn't changed since).
+
+        As long as the timestamp is current enough, we will
+        show the user as active (even if there's a newer
+        timestamp for idle).
+    */
+    if (age(active_timestamp)  < OFFLINE_THRESHOLD_SECS) {
+        return {
+            status: 'active',
+            last_active: active_timestamp,
+        };
+    }
+
+    if (age(idle_timestamp)  < OFFLINE_THRESHOLD_SECS) {
+        return {
+            status: 'idle',
+            last_active: active_timestamp,
+        };
+    }
+
+    return {
+        status: 'offline',
+        last_active: active_timestamp,
```

Comment:
I think the previous code has `last_active = 0` when we don't have any data for the user; this new version seems to use `undefined` for that case. (If the server didn't send an `activet_timestamp`).  Is that change intentional?

Reply:
Yeah, the `last_active = 0` was just a local variable designed to avoid trickiness with JS arithmetic in age comparisons.  The new code no longer has this particular loop, but we have to deal with the same trickiness in other places, hence code like this:

~~~ js
+    function age(timestamp) {
+        return raw.server_timestamp - (timestamp || 0);
+    }
~~~

And this:

~~~ js
+    for (const rec of Object.values(info)) {
+        if (rec.status === 'active') {
+            if (rec.timestamp > (raw.active_timestamp || 0)) {
+                raw.active_timestamp = rec.timestamp;
 ~~~